### PR TITLE
[RFC] emulation devices for t1rocket

### PR DESCRIPTION
**Pull request by OceanS2000**
Add extensible devices emulation for t1rocket simulation dpi backend. Currently only simple Memory device is implemented, and more devices with more functionality are expected to come.

### CODE REVIEW

**Review by FanShupei — CHANGES_REQUESTED** [review]

File: difftest/dpi_t1rocket/src/bus.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        ShadowBusDevice {
+          base: 0xc0000000,
+          size: SRAM_SIZE,
+          device: MemDevice::<SRAM_SIZE>::new(),
+        },
+      ],
+    }
+  }
+
+  pub fn read_mem_axi(&self, addr: u32, size: u32, bus_size: u32) -> Vec<u8> {
```
[the remark this answers is not in the document]
Comment (OceanS2000) [reply difftest/dpi_t1rocket/src/bus.rs:64]: done

File: difftest/dpi_t1rocket/src/bus.rs [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        }
+      }
+      None => {
+        error!("read addr={addr:#x} size={size}B dlen={bus_size}B leads to nowhere!");
+        vec![0; bus_size as usize]
+      }
+    }
+  }
+
+  pub fn write_mem_axi(
```
[the remark this answers is not in the document]
Comment (OceanS2000) [reply difftest/dpi_t1rocket/src/bus.rs:104]: done

File: difftest/dpi_t1rocket/src/bus/mem.rs [hunk 3]
```diff
@@ -0,0 +1,44 @@
+use super::ShadowDevice;
+
+pub(super) struct MemDevice<const SIZE: usize> {
+  mem: [u8; SIZE],
+}
```
[the remark this answers is not in the document]
Comment (OceanS2000) [reply difftest/dpi_t1rocket/src/bus/mem.rs]: indirection it goes, then. This should avoid big stack allocation while still convey that `mem` shall be a fixed-size array.

**Comment by FanShupei** [comment]
LGTM

**Review by sequencer — APPROVED** [review]

**Pull request merged**